File: backend/src/ips_app/presentation/http/handlers/role.py
from typing import List, Optional

from ips_app.domain.models.user import UserAccessTokenClaims
from ips_app.domain.usecases.role import RoleUsecase
from ips_app.presentation.http.dto.common import MessageResponse, PaginatedResponse
from ips_app.presentation.http.dto.permission import PermissionResponse
from ips_app.presentation.http.dto.role import (
    CreateRoleRequest,
    RolePermissionIdsRequest,
    RoleResponse,
    UpdateRolePreferencesRequest,
    UpdateRoleRequest,
)


class RoleHandler:
    def __init__(self, usecase: RoleUsecase) -> None:
        self.usecase = usecase

    async def post_role(
        self,
        request: CreateRoleRequest,
        claims: Optional[UserAccessTokenClaims],
    ) -> RoleResponse:
        role = await self.usecase.create_role(
            name=request.name,
            description=request.description,
            is_default=request.is_default,
            created_by=claims.user_id if claims else None,
        )
        return RoleResponse.from_domain(role)

    async def get_role(self, role_id: str) -> RoleResponse:
        role = await self.usecase.get_role_by_id(role_id)
        return RoleResponse.from_domain(role)

    async def get_default_role(self) -> RoleResponse:
        role = await self.usecase.get_default_role()
        return RoleResponse.from_domain(role)

    async def get_roles(
        self, page: int, limit: int, search: Optional[str]
    ) -> PaginatedResponse[RoleResponse]:
        roles, total = await self.usecase.get_roles(page=page, limit=limit, search=search)
        return PaginatedResponse[RoleResponse](
            items=[RoleResponse.from_domain(r) for r in roles],
            page=page,
            limit=limit,
            total=total,
        )

    async def get_role_permissions(self, role_id: str) -> List[PermissionResponse]:
        permissions = await self.usecase.get_role_permissions(role_id)
        return [PermissionResponse.from_domain(p) for p in permissions]

    async def patch_role(
        self,
        role_id: str,
        request: UpdateRoleRequest,
        claims: Optional[UserAccessTokenClaims],
    ) -> RoleResponse:
        role = await self.usecase.update_role(
            id=role_id,
            name=request.name,
            description=request.description,
            updated_by=claims.user_id if claims else None,
        )
        return RoleResponse.from_domain(role)

    async def patch_default_role(
        self, role_id: str, claims: Optional[UserAccessTokenClaims]
    ) -> RoleResponse:
        role = await self.usecase.set_default_role(
            id=role_id, updated_by=claims.user_id if claims else None
        )
        return RoleResponse.from_domain(role)

    async def patch_role_preferences(
        self,
        role_id: str,
        request: UpdateRolePreferencesRequest,
        claims: Optional[UserAccessTokenClaims],
    ) -> RoleResponse:
        role = await self.usecase.update_role_preferences(
            id=role_id,
            preferences=request.preferences,
            updated_by=claims.user_id if claims else None,
        )
        return RoleResponse.from_domain(role)

    async def delete_role(self, role_id: str) -> MessageResponse:
        await self.usecase.delete_role(role_id)
        return MessageResponse(message="Role deleted successfully.")

    async def post_role_permissions(
        self,
        role_id: str,
        request: RolePermissionIdsRequest,
        claims: Optional[UserAccessTokenClaims],
    ) -> RoleResponse:
        role = await self.usecase.add_permissions_to_role(
            id=role_id,
            permission_ids=list(request.permission_ids),
            updated_by=claims.user_id if claims else None,
        )
        return RoleResponse.from_domain(role)

    async def delete_role_permissions(
        self,
        role_id: str,
        request: RolePermissionIdsRequest,
        claims: Optional[UserAccessTokenClaims],
    ) -> RoleResponse:
        role = await self.usecase.remove_permissions_from_role(
            id=role_id,
            permission_ids=list(request.permission_ids),
            updated_by=claims.user_id if claims else None,
        )
        return RoleResponse.from_domain(role)

File: frontend/src/lib/api/role.ts
import { requestJson, type ApiRequestOptions } from "./client";
import type {
  AuditedFields,
  JsonObject,
  MessageResponse,
  PaginatedResponse,
  PaginationQuery,
  PermissionIdsRequest,
} from "./common";
import type { PermissionResponse } from "./permission";

export type AddRoleRequest = {
  name: string;
  description?: string;
  is_default?: boolean;
};

export type SetRoleRequest = {
  name?: string;
  description?: string;
};

export type RoleResponse = AuditedFields & {
  id: string;
  name: string;
  description: string;
  is_default: boolean;
  permissions: PermissionResponse[];
  preferences: JsonObject;
};

export type RolesResponse = PaginatedResponse<RoleResponse>;

export type GetRolesQuery = PaginationQuery;

export function createRole(
  request: AddRoleRequest,
  options?: ApiRequestOptions,
): Promise<RoleResponse> {
  return requestJson<RoleResponse>("/roles", {
    ...options,
    json: request,
    method: "POST",
  });
}

export function getRoles(
  query: GetRolesQuery = {},
  options?: ApiRequestOptions,
): Promise<RolesResponse> {
  return requestJson<RolesResponse>("/roles", {
    ...options,
    query,
  });
}

export function getDefaultRole(
  options?: ApiRequestOptions,
): Promise<RoleResponse> {
  return requestJson<RoleResponse>("/roles/default", options);
}

export function getRole(
  roleId: string,
  options?: ApiRequestOptions,
): Promise<RoleResponse> {
  return requestJson<RoleResponse>(
    `/roles/${encodeURIComponent(roleId)}`,
    options,
  );
}

export function updateRole(
  roleId: string,
  request: SetRoleRequest,
  options?: ApiRequestOptions,
): Promise<RoleResponse> {
  return requestJson<RoleResponse>(`/roles/${encodeURIComponent(roleId)}`, {
    ...options,
    json: request,
    method: "PATCH",
  });
}

export function setDefaultRole(
  roleId: string,
  options?: ApiRequestOptions,
): Promise<RoleResponse> {
  return requestJson<RoleResponse>(
    `/roles/${encodeURIComponent(roleId)}/default`,
    {
      ...options,
      method: "PATCH",
    },
  );
}

export function updateRolePreferences(
  roleId: string,
  preferences: JsonObject,
  options?: ApiRequestOptions,
): Promise<RoleResponse> {
  return requestJson<RoleResponse>(
    `/roles/${encodeURIComponent(roleId)}/preferences`,
    {
      ...options,
      json: preferences,
      method: "PATCH",
    },
  );
}

export function deleteRole(
  roleId: string,
  options?: ApiRequestOptions,
): Promise<MessageResponse> {
  return requestJson<MessageResponse>(`/roles/${encodeURIComponent(roleId)}`, {
    ...options,
    method: "DELETE",
  });
}

export function addRolePermissions(
  roleId: string,
  request: PermissionIdsRequest,
  options?: ApiRequestOptions,
): Promise<RoleResponse> {
  return requestJson<RoleResponse>(
    `/roles/${encodeURIComponent(roleId)}/permissions`,
    {
      ...options,
      json: request,
      method: "POST",
    },
  );
}

export function removeRolePermissions(
  roleId: string,
  request: PermissionIdsRequest,
  options?: ApiRequestOptions,
): Promise<RoleResponse> {
  return requestJson<RoleResponse>(
    `/roles/${encodeURIComponent(roleId)}/permissions`,
    {
      ...options,
      json: request,
      method: "DELETE",
    },
  );
}

export function getRolePermissions(
  roleId: string,
  options?: ApiRequestOptions,
): Promise<PermissionResponse[]> {
  return requestJson<PermissionResponse[]>(
    `/roles/${encodeURIComponent(roleId)}/permissions`,
    options,
  );
}
